module/os/freebsd/spl/spl_page.c

```diff
@@ -0,0 +1,221 @@
+/*
+ * CDDL HEADER START
+ *
+ * The contents of this file are subject to the terms of the
+ * Common Development and Distribution License (the "License").
+ * You may not use this file except in compliance with the License.
+ *
+ * You can obtain a copy of the license at usr/src/OPENSOLARIS.LICENSE
+ * or http://www.opensolaris.org/os/licensing.
+ * See the License for the specific language governing permissions
+ * and limitations under the License.
+ *
+ * When distributing Covered Code, include this CDDL HEADER in each
+ * file and include the License file at usr/src/OPENSOLARIS.LICENSE.
+ * If applicable, add the following below this CDDL HEADER, with the
+ * fields enclosed by brackets "[]" replaced with your own identifying
+ * information: Portions Copyright [yyyy] [name of copyright owner]
+ *
+ * CDDL HEADER END
+ */
+/*
+ * Copyright 2009 Sun Microsystems, Inc.  All rights reserved.
+ * Use is subject to license terms.
+ */
+
+/* Copyright (c) 1984, 1986, 1987, 1988, 1989 AT&T */
+/* All Rights Reserved   */
+
+/*
+ * University Copyright- Copyright (c) 1982, 1986, 1988
+ * The Regents of the University of California
+ * All Rights Reserved
+ *
+ * University Acknowledgment- Portions of this document are derived from
+ * software developed by the University of California, Berkeley, and its
+ * contributors.
+ */
+
+/*
+ * $FreeBSD$
+ */
+
+#include <sys/page.h>
+#include <sys/byteorder.h>
+#include <sys/lock.h>
+#include <sys/vm.h>
+#include <vm/vm_map.h>
+#include <sys/zfs_context.h>
+
+/*
+ *
+ * if the operation is marked as read then we are stating the kernel will
+ * be writing to pages and must give it write access.
+ *
+ *
+ * This code comes from the FreeBSD's linuxkpi and models their implementation
+ * of get_user_pages(), set_page_dirty(), and put_page().
+ */
+long
+zfs_hold_pages(unsigned long start, unsigned long nr_pages, int read,
+    zfs_page_p *pages)
+{
+	vm_map_t map;
+	vm_prot_t prot;
+	size_t len;
+	int count;
+	int write = read ? 1 : 0;
+
+	map = &curthread->td_proc->p_vmspace->vm_map;
+
```

Comment:
This permission combination doesn't make sense to me. Reading from disk means writing to memory. So the logic should be the other way around unless you've inverted the sense elsewhere, but it looks like you're just passing `cmd != UIO_WRITE`

Reply:
I updated this as well to make it more clear what is going on here.